### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by carleondel**
Notebook ready, Random forests, Gradient boosting Trees and Catboost models explored with different hyperparameters and final comparison and model selection made.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -10,12 +10,14 @@ python = "^3.11"
...
 numpy = "^1.25.2"
-sklearn = "^0.0.post7"
 requests = "^2.31.0"
 black = "^23.7.0"
 flake8 = "^6.1.0"
 mypy = "^1.5.1"
 pytest = "^7.4.0"
+seaborn = "^0.12.2"
+joblib = "^1.3.2"
+scikit-learn = "^1.3.0"
```
Comment (pabaldonedo) [inline pyproject.toml:21]: Are not we missing catboost?

File: src/module_4/models/CatBoost.joblib
Comment (pabaldonedo) [inline src/module_4/models/CatBoost.joblib:1]: Models do not go into the repo as a general rule, but into a model registry

File: src/module_4/push_notifications_project.md [hunk 2]
```diff
@@ -0,0 +1,7169 @@
+## Exploration Phase - Non Linear Models & Model Selection
+
+**GOAL**: Create a Classification Model that can predict whether or not a person would buy an item we offered them (via push notification) based on behavioral and personal features of that user (user id, ordered before, etc), features of that specific order (date, etc) and features of the items themselves (popularity, price, avg days to buy, etc)
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:3]: Keep documenting the start of the notebook, good!

File: src/module_4/push_notifications_project.md [hunk 3]
```diff
@@ -0,0 +1,7169 @@
...
+
+from catboost import Pool
+from catboost import CatBoostClassifier
+
+
+```
+
+
+```python
+path = '/home/carleondel/data-zrive-ds/box_builder_dataset/'
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:52]: Just reminding you the existence of Path from pathlib as recommended waya to deal with paths

File: src/module_4/push_notifications_project.md [hunk 4]
```diff
@@ -0,0 +1,7169 @@
...
+We already explored our dataset so we will jump directly into splitting it
+
+### Data splitting
+- Before splitting we need to filter our df to keep only orders with more than 5 items
+- Then we make a temporal split of 70-20-10 for train | validation | test
+- Finally we make the X | y splits
+
+
+```python
+def push_relevant_dataframe(df: pd.DataFrame, min_products: int=5) -> pd.DataFrame:
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:65]: Good use of methods even within a notebook

File: src/module_4/push_notifications_project.md [hunk 5]
```diff
@@ -0,0 +1,7169 @@
...
+     'avg_days_to_buy_product_type',
+     'std_days_to_buy_product_type',
+     'ordered_before',
+     'abandoned_before',
+     'active_snoozed',
+     'set_as_regular']
+
+
+
+Before creating our baseline model we are going to define some functions that will help us plot the different metrics we will use
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:432]: Nice!

File: src/module_4/push_notifications_project.md [hunk 6]
```diff
@@ -0,0 +1,7169 @@
...
+
+
+```python
+def plot_metrics(
+        model_name : str, y_pred : pd.Series, y_test : pd.Series,
+        figure : Tuple[matplotlib.figure.Figure, np.array] = None
+    ):
+    precision_, recall_, _ = precision_recall_curve(
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:440]: TIP on style, you can use black so the style is consistent and independent of the developer

File: src/module_4/push_notifications_project.md [hunk 7]
```diff
@@ -0,0 +1,7169 @@
...
+    ax[0].set_title("Precision-recall Curve")
+    ax[0].legend()
+
+
+
+    ax[1].plot(fpr, tpr, label=f"AUC: {roc_auc:.2f}")
+    ax[1].set_xlabel("FPR")
+    ax[1].set_ylabel("TPR")
+    ax[1].set_title("ROC Curve")
+    ax[1].legend()
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:466]: return the fig and ax, since they are created here and the user may want to expand the plot further

File: src/module_4/push_notifications_project.md [hunk 8]
```diff
@@ -0,0 +1,7169 @@
...
+```
+
+### Baseline model
+Since we are now exploring non linear models, our baseline model will be the best linear model we got in module 3, which is Ridge Regression with C=1e-6
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:470]: This is true for developing. However, when presenting results to stakeholders always keep the basic baseline too so you can create a story on how getting increasing performance with better models and how much lift you are generating

File: src/module_4/push_notifications_project.md [hunk 9]
```diff
@@ -0,0 +1,7169 @@
...
+      <td>25.0</td>
+      <td>21.04899</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+We only have 2 instances classified as 1 with probabilities >10%. So we should consider lowering the threshold
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:604]: Remember, these are not probabilities but scores. Nothing ensures you that 10% of the times with that scoring the label is 1

Comment (pabaldonedo) [reply src/module_4/push_notifications_project.md:604]: Also it is not clear to me where this 10% comes from, was the selected one in previous module? Why it was like that before?

File: src/module_4/push_notifications_project.md [hunk 10]
```diff
@@ -0,0 +1,7169 @@
...
+print(f'Recall: {recall:.2f}')
+```
+
+    Performance with new threshold (0.02):
+    Accuracy: 0.98
+    Precision: 0.25
+    Recall: 0.22
+
+
+In our context it makes more sense to accept more false positive cases, so we could choose a threshold between 0.03 and 0.02.
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:662]: A couple of things here:
 - The way it is presented it seems that you have first chosen the thresholds and then check the metrics. You should create the story of checking the plot we select the point with a precision-recall around X. And then we check to which threshold it corresponds.
 - I would advocate to first compare models using the curves and picking the working threshold as last step for the winning model

File: src/module_4/push_notifications_project.md [hunk 11]
```diff
@@ -0,0 +1,7169 @@
...
+```python
+# We initialize the subplots
+fig1, ax1 = plt.subplots(1, 2, figsize=(14,7))
+fig1.suptitle("Train metrics")
+
+fig2, ax2 = plt.subplots(1, 2, figsize = (14,7))
+fig2.suptitle("Validation metrics")
+
+# Number of trees
+# In random forests we don't need to scale our training data
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:679]: Nice that you remember this :D

File: src/module_4/push_notifications_project.md [hunk 12]
```diff
@@ -0,0 +1,7169 @@
...
+
+
+
+    
+![png](push_notifications_project_files/push_notifications_project_32_1.png)
+    
+
+
+### Insights 
+- At first glance, we can clearly see how this model works. The performance metrics are much better on train than on validation (overfit signal). This is because each tree is built in its entirety, making the model learn some noise. But because we are averaging the predictions from multiple trees, this overfitting is reduced.
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:712]: This is not true. By construction training error in RF is 0 (unless two samples are identical with different label), which does not mean it is overfitting. Check the book elements of statical learning or review the class or slack me for the full details.

What could be happening:
 - A lot of correlation among trees
 - Too many low power features makes some trees to receive very poor feature sets making them to have very high bias increasing the total model bias

File: src/module_4/push_notifications_project.md [hunk 13]
```diff
@@ -0,0 +1,7169 @@
...
+- After creating different models using different number of trees, there is not much improvement when going from 50 to 100, so it doesn't make sense to keep training with more trees.
+- Overall the performance of our Random Forest models is slightly worse than the performance of our Baseline Model (Ridge C=1e-6)
+
+### Feature Importance & Retraining
+Now we are going to reduce our features and retrain our model to see if we make any improvement. Since there wasn't much improvement when moving from 50 to 100 trees, we will base our decisions on our model with 50 trees.
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:717]: What is the motivation for doing this?

File: src/module_4/push_notifications_project.md [hunk 14]
```diff
@@ -0,0 +1,7169 @@
...
+      <td>count_babies</td>
+      <td>0.001029</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+We are going to keep the top 3 features by importance, train again and compare
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:873]: Why only 3 and not 2 or 5 or 10?

File: src/module_4/push_notifications_project.md [hunk 15]
```diff
@@ -0,0 +1,7169 @@
...
+
+
+
+    
+![png](push_notifications_project_files/push_notifications_project_39_1.png)
+    
+
+
+### Insights
+- The performance metrics are significantly worse than the metrics when training with all the features. (PR-AUC = 0.04 vs 0.13)
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:923]: I think this part has not been properly thought through and the decision are not well justified as I state via the questions I have left.

File: src/module_4/push_notifications_project.md [hunk 16]
```diff
@@ -0,0 +1,7169 @@
...
+
+
+
+    
+![png](push_notifications_project_files/push_notifications_project_43_1.png)
+    
+
+
+## Insights
+This took >20 mins training...
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:1000]: This could be due to too large dataset or very low training rate. You can try different learning rates to check the results. Also if you have watched already module 5 class you have learned about the learning curves that can give you hint if your learning rate is high or low

File: src/module_4/push_notifications_project.md [hunk 17]
```diff
@@ -0,0 +1,7169 @@
...
+
+Our Baseline model had PR-AUC = 0.16, ROC-AUC = 0.83. So we have finally made some improvement.
+- When we used max_depth of 5, all models had a similar performance, independently of the other hyperparameter (number of estimators). PR-AUC = 0.22 and ROC-AUC = 0.79.
+- So our model selected from Gradient Boosting Trees will be the one with hyperparameters: Number of estimators = 100, max_depth = 5
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:1004]: 100 trees and depth 1 has better PR-AUC why not choosing that?
In addition I find interesting that the results are the same in training regardless of the number of estimators, like you are hitting the minima super early. Probably reducing learning rate can help here.

File: src/module_4/push_notifications_project.md [hunk 18]
```diff
@@ -0,0 +1,7169 @@
...
+    fpr, tpr, _ = roc_curve(y_test, y_pred)
+    roc_auc = auc(fpr, tpr)
+    precision_, recall_, _ = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall_, precision_)
+    return roc_auc, pr_auc
+```
+
+
+```python
+# We are going to create a grid search based on my criteria
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:1093]: Can also consider random search, specially if you have performance (in the sense of training time) issues

File: src/module_4/push_notifications_project.md [hunk 19]
```diff
@@ -0,0 +1,7169 @@
...
+            hyper_lr.append(lr)
+
+            n = n+1
+            # Actualizations of the training
+            print(
+                f"Trained model nº {n}/27. Iterations: {iterations}"
+                f"Depth: {depth}"
+                f"Learning rate: {lr}"
+            )
+
```
Comment (pabaldonedo) [inline src/module_4/push_notifications_project.md:1137]: Given that this is a learning exercise, it may be worth investing some time extracting insights on how different parametrizations affect training and test metrics, overfit and underfit

**Review by pabaldonedo — COMMENTED** [review]
Very nice work in general.
Some comments:
- CRITICAL! revisit the Random Forest theory (book, class, slack...) since you made wrong statements that indicate you should go over it again.
- I missed a final decision on the models: choose the best one, picking the threshold and evaluate the final metrics (precision-recall)
- I would try to extract insights from the different parametrizations about under and overfit
- I would explore more learning rates for GBT and try to understand how it is affecting

If you happen to work more on this (xgboost or productivisation) just re-request a review.
